File: backend/simulateur.py
# =============================================================================
# simulateur.py — Simulateur de données capteurs ESP32
# Simule des cas réels de température pour tester l'IA
# Exécuter : python simulateur.py
# =============================================================================

import requests
import time
import random
import math
from datetime import datetime

# -----------------------------------------------------------------------------
# CONFIGURATION
# Modifier l'URL si le backend tourne sur un autre port
# Modifier l'equipement_id selon ce que tu as créé dans l'app
# -----------------------------------------------------------------------------
API_URL       = "http://localhost:8000"
EQUIPEMENT_ID = 1        # ID de l'équipement dans la base
INTERVALLE    = 3        # Secondes entre chaque mesure (3s pour voir vite)

# -----------------------------------------------------------------------------
# COULEURS pour le terminal
# -----------------------------------------------------------------------------
VERT    = "\033[92m"
ORANGE  = "\033[93m"
ROUGE   = "\033[91m"
BLEU    = "\033[94m"
RESET   = "\033[0m"
GRAS    = "\033[1m"

def couleur_temp(temp):
    if temp >= 85: return ROUGE
    if temp >= 70: return ORANGE
    return VERT

def envoyer_mesure(valeur):
    """Envoie une mesure au backend FastAPI"""
    try:
        response = requests.post(
            f"{API_URL}/api/capteurs/mesure",
            json={
                "type_capteur":  "temperature",
                "valeur":        round(valeur, 2),
                "unite":         "°C",
                "equipement_id": EQUIPEMENT_ID
            },
            timeout=5
        )
        return response.status_code == 201
    except requests.exceptions.ConnectionError:
        print(f"{ROUGE}❌ Backend non accessible — vérifiez que uvicorn tourne{RESET}")
        return False

def analyser_avec_ia():
    """Déclenche l'analyse IA après envoi de mesures"""
    try:
        response = requests.post(
            f"{API_URL}/api/ia/analyser/{EQUIPEMENT_ID}",
            timeout=5
        )
        if response.status_code == 200:
            data = response.json()
            statut = data.get("statut", "normal")
            proba  = data.get("probabilite", 0)
            if statut == "critique":
                print(f"  {ROUGE}{GRAS}🚨 IA : CRITIQUE — Probabilité panne {int(proba*100)}% — OT généré automatiquement !{RESET}")
            elif statut == "warning":
                print(f"  {ORANGE}⚠  IA : AVERTISSEMENT — Probabilité panne {int(proba*100)}%{RESET}")
            else:
                print(f"  {VERT}✅ IA : Normal{RESET}")
    except:
        pass

def barre_progression(valeur, max_val=100, largeur=30):
    """Affiche une barre de progression colorée"""
    pct     = min(valeur / max_val, 1.0)
    rempli  = int(pct * largeur)
    vide    = largeur - rempli
    couleur = couleur_temp(valeur)
    return f"{couleur}{'█' * rempli}{'░' * vide}{RESET} {couleur}{valeur:.1f}°C{RESET}"

# =============================================================================
# SCÉNARIOS DE SIMULATION
# =============================================================================

def scenario_normal(duree=30):
    """
    Scénario 1 : Fonctionnement normal
    Température stable entre 55-65°C avec légères variations
    """
    print(f"\n{BLEU}{GRAS}═══ SCÉNARIO 1 : Fonctionnement Normal ═══{RESET}")
    print(f"Température stable entre 55-65°C — {duree} mesures\n")

    base = 60.0
    for i in range(duree):
        # Variation aléatoire naturelle
        bruit    = random.gauss(0, 1.5)
        sinusoide = 3 * math.sin(i * 0.3)
        temp     = base + bruit + sinusoide
        temp     = max(50, min(68, temp))

        ok = envoyer_mesure(temp)
        statut_envoi = f"{VERT}✓{RESET}" if ok else f"{ROUGE}✗{RESET}"

        print(f"  [{i+1:02d}/{duree}] {statut_envoi} {barre_progression(temp)} "
              f"| {datetime.now().strftime('%H:%M:%S')}")

        if (i + 1) % 10 == 0:
            analyser_avec_ia()

        time.sleep(INTERVALLE)


def scenario_montee_progressive(duree=40):
    """
    Scénario 2 : Montée progressive (usure, surcharge)
    Température monte de 60°C à 80°C sur la durée
    """
    print(f"\n{ORANGE}{GRAS}═══ SCÉNARIO 2 : Montée Progressive (Usure) ═══{RESET}")
    print(f"Température montant de 60°C à 80°C — {duree} mesures\n")

    for i in range(duree):
        progression = i / duree
        temp_base   = 60 + (progression * 20)
        bruit       = random.gauss(0, 0.8)
        temp        = temp_base + bruit

        ok = envoyer_mesure(temp)
        statut_envoi = f"{VERT}✓{RESET}" if ok else f"{ROUGE}✗{RESET}"

        print(f"  [{i+1:02d}/{duree}] {statut_envoi} {barre_progression(temp)} "
              f"| Tendance: +{progression*20:.1f}°C | {datetime.now().strftime('%H:%M:%S')}")

        if (i + 1) % 8 == 0:
            analyser_avec_ia()

        time.sleep(INTERVALLE)


def scenario_surchauffe_critique(duree=25):
    """
    Scénario 3 : Surchauffe critique (panne imminente)
    Température dépasse 85°C — l'IA doit générer un OT automatiquement
    """
    print(f"\n{ROUGE}{GRAS}═══ SCÉNARIO 3 : SURCHAUFFE CRITIQUE 🔥 ═══{RESET}")
    print(f"Température dépassant 85°C — L'IA va générer un OT automatique !\n")

    for i in range(duree):
        if i < 8:
            # Montée rapide
            temp = 70 + (i * 2.5) + random.gauss(0, 0.5)
        else:
            # Zone critique avec oscillations
            temp = 88 + random.gauss(0, 2)
            temp = max(84, min(96, temp))

        ok = envoyer_mesure(temp)
        statut_envoi = f"{VERT}✓{RESET}" if ok else f"{ROUGE}✗{RESET}"

        alerte = f"{ROUGE}{GRAS} ⚠ ZONE CRITIQUE !{RESET}" if temp >= 85 else ""
        print(f"  [{i+1:02d}/{duree}] {statut_envoi} {barre_progression(temp)}{alerte} "
              f"| {datetime.now().strftime('%H:%M:%S')}")

        if (i + 1) % 5 == 0:
            analyser_avec_ia()

        time.sleep(INTERVALLE)


def scenario_retour_normal(duree=20):
    """
    Scénario 4 : Retour à la normale après intervention
    Simule le refroidissement après une maintenance
    """
    print(f"\n{VERT}{GRAS}═══ SCÉNARIO 4 : Retour Normal (Après Maintenance) ═══{RESET}")
    print(f"Refroidissement de 88°C à 58°C — {duree} mesures\n")

    for i in range(duree):
        progression = i / duree
        temp_base   = 88 - (progression * 30)
        bruit       = random.gauss(0, 0.6)
        temp        = max(55, temp_base + bruit)

        ok = envoyer_mesure(temp)
        statut_envoi = f"{VERT}✓{RESET}" if ok else f"{ROUGE}✗{RESET}"

        print(f"  [{i+1:02d}/{duree}] {statut_envoi} {barre_progression(temp)} "
              f"| Refroidissement: -{progression*30:.1f}°C | {datetime.now().strftime('%H:%M:%S')}")

        if (i + 1) % 7 == 0:
            analyser_avec_ia()

        time.sleep(INTERVALLE)


def scenario_pic_soudain(duree=30):
    """
    Scénario 5 : Pic soudain (court-circuit ou blocage mécanique)
    Température normale puis pic brutal suivi de retour
    """
    print(f"\n{ROUGE}{GRAS}═══ SCÉNARIO 5 : Pic Soudain (Court-circuit) ═══{RESET}")
    print(f"Pic brutal à 92°C puis retour — {duree} mesures\n")

    for i in range(duree):
        if i < 10:
            # Phase normale
            temp = 62 + random.gauss(0, 1.5)
        elif i < 15:
            # Pic brutal
            temp = 62 + ((i - 10) / 5) * 30 + random.gauss(0, 1)
        elif i < 20:
            # Zone critique
            temp = 90 + random.gauss(0, 2)
        else:
            # Retour progressif
            temp = 90 - ((i - 20) / 10) * 30 + random.gauss(0, 1.5)
            temp = max(58, temp)

        ok = envoyer_mesure(temp)
        statut_envoi = f"{VERT}✓{RESET}" if ok else f"{ROUGE}✗{RESET}"

        phase = "Normal" if i < 10 else ("↑ Montée" if i < 15 else ("🔥 Pic" if i < 20 else "↓ Retour"))
        print(f"  [{i+1:02d}/{duree}] {statut_envoi} {barre_progression(temp)} "
              f"| Phase: {phase} | {datetime.now().strftime('%H:%M:%S')}")

        if (i + 1) % 5 == 0:
            analyser_avec_ia()

        time.sleep(INTERVALLE)


def scenario_complet():
    """
    Scénario COMPLET : Enchaîne tous les scénarios
    Parfait pour une démonstration complète en soutenance
    """
    print(f"\n{BLEU}{GRAS}{'='*55}{RESET}")
    print(f"{BLEU}{GRAS}   DÉMONSTRATION COMPLÈTE GMAO — Maintenance 4.0{RESET}")
    print(f"{BLEU}{GRAS}{'='*55}{RESET}")
    print(f"  Équipement ID : {EQUIPEMENT_ID}")
    print(f"  Intervalle    : {INTERVALLE}s entre chaque mesure")
    print(f"  Backend       : {API_URL}")
    print(f"{BLEU}{GRAS}{'='*55}{RESET}\n")

    scenario_normal(duree=20)
    print(f"\n{GRAS}⏳ Pause 3 secondes...{RESET}\n")
    time.sleep(3)

    scenario_montee_progressive(duree=25)
    print(f"\n{GRAS}⏳ Pause 3 secondes...{RESET}\n")
    time.sleep(3)

    scenario_surchauffe_critique(duree=20)
    print(f"\n{GRAS}⏳ Pause 3 secondes...{RESET}\n")
    time.sleep(3)

    scenario_retour_normal(duree=15)

    print(f"\n{VERT}{GRAS}{'='*55}{RESET}")
    print(f"{VERT}{GRAS}   ✅ DÉMONSTRATION TERMINÉE{RESET}")
    print(f"{VERT}{GRAS}{'='*55}{RESET}")
    print(f"  → Vérifiez les Alertes IA dans l'application")
    print(f"  → Vérifiez les OT générés automatiquement")
    print(f"  → Vérifiez le graphique dans Capteurs IoT\n")


# =============================================================================
# MENU PRINCIPAL
# =============================================================================
def menu():
    print(f"\n{BLEU}{GRAS}╔══════════════════════════════════════╗{RESET}")
    print(f"{BLEU}{GRAS}║   SIMULATEUR GMAO — Maintenance 4.0  ║{RESET}")
    print(f"{BLEU}{GRAS}╚══════════════════════════════════════╝{RESET}\n")
    print(f"  {GRAS}Équipement ID : {EQUIPEMENT_ID}{RESET}")
    print(f"  {GRAS}Backend       : {API_URL}{RESET}\n")
    print("  Choisissez un scénario :\n")
    print(f"  {VERT}[1]{RESET} Fonctionnement Normal (55-65°C)")
    print(f"  {ORANGE}[2]{RESET} Montée Progressive — Usure (60→80°C)")
    print(f"  {ROUGE}[3]{RESET} Surchauffe Critique 🔥 (>85°C) — OT Auto")
    print(f"  {VERT}[4]{RESET} Retour Normal après Maintenance")
    print(f"  {ROUGE}[5]{RESET} Pic Soudain — Court-circuit")
    print(f"  {BLEU}[6]{RESET} DÉMONSTRATION COMPLÈTE (tous les scénarios)")
    print(f"  [0] Quitter\n")

    choix = input("  Votre choix : ").strip()

    if   choix == "1": scenario_normal()
    elif choix == "2": scenario_montee_progressive()
    elif choix == "3": scenario_surchauffe_critique()
    elif choix == "4": scenario_retour_normal()
    elif choix == "5": scenario_pic_soudain()
    elif choix == "6": scenario_complet()
    elif choix == "0":
        print(f"\n{VERT}Au revoir !{RESET}\n")
        return
    else:
        print(f"{ORANGE}Choix invalide{RESET}")

    input(f"\n{GRAS}Appuyez sur Entrée pour revenir au menu...{RESET}")
    menu()

if __name__ == "__main__":
    menu()

File: frontend/src/pages/Simulateur.js
// =============================================================================
// src/pages/Simulateur.js — Simulateur de données capteurs dans l'application
// Permet de lancer des scénarios réels simulés directement depuis le navigateur
// =============================================================================

import { useState, useRef, useEffect } from 'react';
import { equipementsAPI, capteursAPI, iaAPI } from '../services/api';
import './Simulateur.css';

// ─── Scénarios disponibles ────────────────────────────────────────────────────
const SCENARIOS = [
  {
    id:          1,
    nom:         'Fonctionnement Normal',
    description: 'Température stable entre 55–65°C avec légères variations naturelles',
    icone:       '✅',
    couleur:     'var(--accent-green)',
    duree:       20,
    niveau:      'normal',
  },
  {
    id:          2,
    nom:         'Montée Progressive',
    description: 'Température monte de 60°C à 80°C — simule une usure ou surcharge',
    icone:       '📈',
    couleur:     'var(--accent-orange)',
    duree:       25,
    niveau:      'warning',
  },
  {
    id:          3,
    nom:         'Surchauffe Critique 🔥',
    description: 'Température dépasse 85°C — l\'IA génère un OT automatiquement !',
    icone:       '🚨',
    couleur:     'var(--accent-red)',
    duree:       20,
    niveau:      'critique',
  },
  {
    id:          4,
    nom:         'Retour Normal',
    description: 'Refroidissement de 88°C à 58°C — simule un retour après maintenance',
    icone:       '❄️',
    couleur:     'var(--accent-blue)',
    duree:       15,
    niveau:      'normal',
  },
  {
    id:          5,
    nom:         'Pic Soudain',
    description: 'Température normale puis pic brutal à 92°C — simule un court-circuit',
    icone:       '⚡',
    couleur:     'var(--accent-red)',
    duree:       30,
    niveau:      'critique',
  },
  {
    id:          6,
    nom:         'Démonstration Complète',
    description: 'Enchaîne tous les scénarios — parfait pour la soutenance',
    icone:       '🎓',
    couleur:     'var(--accent-purple)',
    duree:       80,
    niveau:      'complet',
  },
];

// ─── Génération des températures selon le scénario ───────────────────────────
function genererTemperature(scenarioId, index, total) {
  const rand  = () => (Math.random() - 0.5) * 3;
  const gauss = (moy, sigma) => moy + (Math.random() - 0.5) * sigma * 2;

  switch (scenarioId) {
    case 1: // Normal
      return Math.max(50, Math.min(68,
        60 + gauss(0, 1.5) + 3 * Math.sin(index * 0.3)
      ));

    case 2: // Montée progressive
      return 60 + (index / total) * 20 + gauss(0, 0.8);

    case 3: // Surchauffe critique
      if (index < 8) return 70 + index * 2.5 + rand();
      return Math.max(84, Math.min(96, 88 + gauss(0, 2)));

    case 4: // Retour normal
      return Math.max(55, 88 - (index / total) * 30 + gauss(0, 0.6));

    case 5: // Pic soudain
      if (index < 10) return 62 + gauss(0, 1.5);
      if (index < 15) return 62 + ((index - 10) / 5) * 30 + rand();
      if (index < 20) return Math.max(84, Math.min(96, 90 + gauss(0, 2)));
      return Math.max(58, 90 - ((index - 20) / 10) * 30 + gauss(0, 1.5));

    default:
      return 60 + gauss(0, 2);
  }
}

// ─── Composant carte scénario ─────────────────────────────────────────────────
function ScenarioCard({ scenario, actif, enCours, onClick }) {
  return (
    <button
      className={`scenario-card ${actif ? 'actif' : ''} ${enCours && !actif ? 'desactive' : ''}`}
      style={{ '--scenario-couleur': scenario.couleur }}
      onClick={() => onClick(scenario)}
      disabled={enCours}
    >
      <div className="scenario-icone">{scenario.icone}</div>
      <div className="scenario-info">
        <div className="scenario-nom">{scenario.nom}</div>
        <div className="scenario-desc">{scenario.description}</div>
        <div className="scenario-meta">
          <span className={`badge badge-${scenario.niveau === 'normal' ? 'en_service' : scenario.niveau === 'warning' ? 'warning' : scenario.niveau === 'critique' ? 'critique' : 'info'}`}>
            {scenario.niveau}
          </span>
          <span className="scenario-duree">~{scenario.duree} mesures</span>
        </div>
      </div>
    </button>
  );
}

// ─── Page principale Simulateur ───────────────────────────────────────────────
export default function Simulateur() {
  const [equipements, setEquipements]       = useState([]);
  const [equipementId, setEquipementId]     = useState('');
  const [scenarioActif, setScenarioActif]   = useState(null);
  const [enCours, setEnCours]               = useState(false);
  const [logs, setLogs]                     = useState([]);
  const [progression, setProgression]       = useState(0);
  const [total, setTotal]                   = useState(0);
  const [tempActuelle, setTempActuelle]     = useState(null);
  const [stats, setStats]                   = useState({ envoyes: 0, alertes: 0, ot: 0 });
  const stopRef   = useRef(false);
  const logsRef   = useRef(null);

  useEffect(() => {
    equipementsAPI.lister().then(res => {
      setEquipements(res.data);
      if (res.data.length > 0) setEquipementId(res.data[0].id);
    });
  }, []);

  // Auto-scroll logs
  useEffect(() => {
    if (logsRef.current) {
      logsRef.current.scrollTop = logsRef.current.scrollHeight;
    }
  }, [logs]);

  const ajouterLog = (message, type = 'info') => {
    const heure = new Date().toLocaleTimeString('fr-FR');
    setLogs(prev => [...prev, { message, type, heure, id: Date.now() }]);
  };

  const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

  const lancerScenario = async (scenario) => {
    if (!equipementId) {
      ajouterLog('⚠ Sélectionnez un équipement d\'abord', 'warning');
      return;
    }

    // Scénario complet — enchaîne les autres
    if (scenario.id === 6) {
      await lancerDemoComplete();
      return;
    }

    setScenarioActif(scenario);
    setEnCours(true);
    stopRef.current = false;
    setLogs([]);
    setProgression(0);
    setTotal(scenario.duree);
    setStats({ envoyes: 0, alertes: 0, ot: 0 });

    ajouterLog(`🚀 Démarrage : ${scenario.nom}`, 'success');
    ajouterLog(`📡 Équipement ID : ${equipementId} — ${scenario.duree} mesures`, 'info');

    let envoyes = 0, alertes = 0, ot = 0;

    for (let i = 0; i < scenario.duree; i++) {
      if (stopRef.current) {
        ajouterLog('⏹ Simulation arrêtée par l\'utilisateur', 'warning');
        break;
      }

      const temp = genererTemperature(scenario.id, i, scenario.duree);
      setTempActuelle(temp);

      try {
        await capteursAPI.historique; // ping
        const res = await fetch('http://localhost:8000/api/capteurs/mesure', {
          method:  'POST',
          headers: { 'Content-Type': 'application/json' },
          body:    JSON.stringify({
            type_capteur:  'temperature',
            valeur:        parseFloat(temp.toFixed(2)),
            unite:         '°C',
            equipement_id: parseInt(equipementId),
          }),
        });

        if (res.ok) {
          envoyes++;
          const couleurTemp = temp >= 85 ? 'error' : temp >= 70 ? 'warning' : 'success';
          ajouterLog(
            `[${String(i+1).padStart(2,'0')}/${scenario.duree}] 🌡 ${temp.toFixed(1)}°C${temp >= 85 ? ' 🔥 CRITIQUE' : temp >= 70 ? ' ⚠ Élevée' : ''}`,
            couleurTemp
          );
        }
      } catch {
        ajouterLog(`[${i+1}] ❌ Erreur d'envoi`, 'error');
      }

      setProgression(i + 1);
      setStats({ envoyes, alertes, ot });

      // Analyse IA tous les 5 mesures
      if ((i + 1) % 5 === 0) {
        try {
          const iaRes = await fetch(`http://localhost:8000/api/ia/analyser/${equipementId}`, {
            method: 'POST'
          });
          if (iaRes.ok) {
            const data = await iaRes.json();
            if (data.statut === 'critique') {
              alertes++;
              ot += data.ot_genere_id ? 1 : 0;
              ajouterLog(`🤖 IA → CRITIQUE ! Probabilité panne : ${Math.round(data.probabilite * 100)}%${data.ot_genere_id ? ` — OT #${data.ot_genere_id} créé automatiquement !` : ''}`, 'error');
            } else if (data.statut === 'warning') {
              alertes++;
              ajouterLog(`🤖 IA → Avertissement — Probabilité : ${Math.round(data.probabilite * 100)}%`, 'warning');
            } else if (data.statut === 'normal') {
              ajouterLog(`🤖 IA → Normal ✓`, 'success');
            }
            setStats({ envoyes, alertes, ot });
          }
        } catch { /* ignore */ }
      }

      await sleep(1500); // 1.5 secondes entre chaque mesure
    }

    ajouterLog(`✅ Simulation terminée — ${envoyes} mesures envoyées`, 'success');
    setEnCours(false);
    setScenarioActif(null);
  };

  const lancerDemoComplete = async () => {
    const sequence = [
      { ...SCENARIOS[0], duree: 15 },
      { ...SCENARIOS[1], duree: 15 },
      { ...SCENARIOS[2], duree: 15 },
      { ...SCENARIOS[3], duree: 12 },
    ];

    setEnCours(true);
    stopRef.current = false;
    setLogs([]);
    setStats({ envoyes: 0, alertes: 0, ot: 0 });

    ajouterLog('🎓 DÉMONSTRATION COMPLÈTE — Maintenance 4.0', 'success');
    ajouterLog('Enchaînement de 4 scénarios...', 'info');

    for (const sc of sequence) {
      if (stopRef.current) break;
      setScenarioActif(sc);
      setTotal(sc.duree);
      setProgression(0);
      ajouterLog(`\n━━━ ${sc.icone} ${sc.nom} ━━━`, 'info');
      await lancerScenarioInterne(sc);
      if (!stopRef.current) {
        ajouterLog('⏳ Pause 2 secondes...', 'info');
        await sleep(2000);
      }
    }

    ajouterLog('\n🎉 DÉMONSTRATION TERMINÉE !', 'success');
    ajouterLog('→ Vérifiez Capteurs IoT pour le graphique', 'info');
    ajouterLog('→ Vérifiez Alertes IA pour les détections', 'info');
    ajouterLog('→ Vérifiez Ordres de Travail pour les OT générés', 'info');
    setEnCours(false);
    setScenarioActif(null);
  };

  const lancerScenarioInterne = async (scenario) => {
    for (let i = 0; i < scenario.duree; i++) {
      if (stopRef.current) break;

      const temp = genererTemperature(scenario.id, i, scenario.duree);
      setTempActuelle(temp);

      try {
        await fetch('http://localhost:8000/api/capteurs/mesure', {
          method:  'POST',
          headers: { 'Content-Type': 'application/json' },
          body:    JSON.stringify({
            type_capteur:  'temperature',
            valeur:        parseFloat(temp.toFixed(2)),
            unite:         '°C',
            equipement_id: parseInt(equipementId),
          }),
        });

        const couleur = temp >= 85 ? 'error' : temp >= 70 ? 'warning' : 'success';
        ajouterLog(
          `  [${String(i+1).padStart(2,'0')}] 🌡 ${temp.toFixed(1)}°C${temp >= 85 ? ' 🔥' : temp >= 70 ? ' ⚠' : ''}`,
          couleur
        );
      } catch { /* ignore */ }

      setProgression(i + 1);

      if ((i + 1) % 5 === 0) {
        try {
          const iaRes = await fetch(`http://localhost:8000/api/ia/analyser/${equipementId}`, { method: 'POST' });
          if (iaRes.ok) {
            const data = await iaRes.json();
            if (data.statut !== 'normal' && data.probabilite) {
              const type = data.statut === 'critique' ? 'error' : 'warning';
              ajouterLog(`  🤖 IA → ${data.statut.toUpperCase()} (${Math.round(data.probabilite*100)}%)${data.ot_genere_id ? ' — OT généré !' : ''}`, type);
            }
          }
        } catch { /* ignore */ }
      }

      await sleep(1500);
    }
  };

  const arreter = () => {
    stopRef.current = true;
    ajouterLog('⏹ Arrêt demandé...', 'warning');
  };

  const getTempColor = (t) => {
    if (!t) return 'var(--text-muted)';
    if (t >= 85) return 'var(--accent-red)';
    if (t >= 70) return 'var(--accent-orange)';
    return 'var(--accent-green)';
  };

  const pct = total > 0 ? Math.round((progression / total) * 100) : 0;

  return (
    <div className="page-content fade-in">

      <div className="page-header">
        <div>
          <h1 className="page-title">Simulateur IoT</h1>
          <p className="page-subtitle">Simulation de données capteurs en temps réel — Maintenance 4.0</p>
        </div>
        {enCours && (
          <button className="btn btn-danger" onClick={arreter}>⏹ Arrêter</button>
        )}
      </div>

      <div className="sim-layout">

        {/* Colonne gauche */}
        <div className="sim-left">

          {/* Sélecteur équipement */}
          <div className="card sim-eq-card">
            <h3 className="card-title" style={{ marginBottom: 12 }}>Équipement cible</h3>
            <select className="input" value={equipementId}
              onChange={e => setEquipementId(e.target.value)} disabled={enCours}>
              <option value="">Sélectionner un équipement...</option>
              {equipements.map(eq => (
                <option key={eq.id} value={eq.id}>{eq.code} — {eq.nom}</option>
              ))}
            </select>
            {equipements.length === 0 && (
              <p className="sim-hint">⚠ Créez d'abord un équipement dans la page Équipements</p>
            )}
          </div>

          {/* Scénarios */}
          <div className="sim-scenarios">
            {SCENARIOS.map(sc => (
              <ScenarioCard
                key={sc.id}
                scenario={sc}
                actif={scenarioActif?.id === sc.id}
                enCours={enCours}
                onClick={lancerScenario}
              />
            ))}
          </div>

        </div>

        {/* Colonne droite */}
        <div className="sim-right">

          {/* Température actuelle */}
          <div className="card sim-temp-card">
            <div className="sim-temp-header">
              <div>
                <p className="sim-temp-label">Température envoyée</p>
                <div className="sim-temp-valeur" style={{ color: getTempColor(tempActuelle) }}>
                  {tempActuelle ? `${tempActuelle.toFixed(1)}°C` : '—'}
                </div>
              </div>
              <div className="sim-temp-statut">
                {tempActuelle >= 85 ? (
                  <span className="badge badge-critique">CRITIQUE 🔥</span>
                ) : tempActuelle >= 70 ? (
                  <span className="badge badge-warning">ÉLEVÉE ⚠</span>
                ) : tempActuelle ? (
                  <span className="badge badge-en_service">NORMALE ✓</span>
                ) : (
                  <span className="badge" style={{ background: 'var(--bg-secondary)', color: 'var(--text-muted)' }}>EN ATTENTE</span>
                )}
              </div>
            </div>

            {/* Barre de progression */}
            {enCours && (
              <div className="sim-progress-container">
                <div className="sim-progress-header">
                  <span className="sim-progress-label">
                    {scenarioActif?.nom} — {progression}/{total} mesures
                  </span>
                  <span className="sim-progress-pct">{pct}%</span>
                </div>
                <div className="sim-progress-bar">
                  <div
                    className="sim-progress-fill"
                    style={{
                      width: `${pct}%`,
                      background: scenarioActif?.couleur || 'var(--accent-blue)',
                    }}
                  />
                </div>
              </div>
            )}
          </div>

          {/* Stats */}
          <div className="sim-stats">
            <div className="card sim-stat">
              <div className="sim-stat-val" style={{ color: 'var(--accent-blue)' }}>{stats.envoyes}</div>
              <div className="sim-stat-label">Mesures envoyées</div>
            </div>
            <div className="card sim-stat">
              <div className="sim-stat-val" style={{ color: 'var(--accent-orange)' }}>{stats.alertes}</div>
              <div className="sim-stat-label">Alertes IA</div>
            </div>
            <div className="card sim-stat">
              <div className="sim-stat-val" style={{ color: 'var(--accent-green)' }}>{stats.ot}</div>
              <div className="sim-stat-label">OT générés</div>
            </div>
          </div>

          {/* Logs */}
          <div className="card sim-logs-card">
            <div className="sim-logs-header">
              <h3 className="card-title">Journal en temps réel</h3>
              <button className="btn btn-secondary btn-sm"
                onClick={() => setLogs([])}>Effacer</button>
            </div>
            <div className="sim-logs" ref={logsRef}>
              {logs.length === 0 ? (
                <p className="sim-logs-empty">Lancez un scénario pour voir les logs...</p>
              ) : (
                logs.map(log => (
                  <div key={log.id} className={`sim-log sim-log-${log.type}`}>
                    <span className="sim-log-heure">{log.heure}</span>
                    <span className="sim-log-msg">{log.message}</span>
                  </div>
                ))
              )}
            </div>
          </div>

        </div>
      </div>
    </div>
  );
}
